File: packages/provider-router/src/adapters/openai.ts
import { Provider, ProviderError } from '@vocaliq/shared';
import OpenAI from 'openai';
import type { CompletionOptions, CompletionResult, LLMMessage, LLMProvider } from '../index.js';

/**
 * OpenAI LLM adapter (Chat Completions + Embeddings). Cost-efficient default
 * `gpt-4o-mini`; embeddings default `text-embedding-3-small`. Key injected only.
 */
export class OpenAILLM implements LLMProvider {
  readonly provider = Provider.OPENAI;
  readonly capability = 'llm' as const;
  readonly defaultModel = 'gpt-4o-mini';
  readonly defaultEmbeddingModel = 'text-embedding-3-small';
  private readonly client: OpenAI;

  constructor(apiKey: string) {
    this.client = new OpenAI({ apiKey });
  }

  private toMessages(
    messages: LLMMessage[],
    opts?: CompletionOptions,
  ): OpenAI.Chat.ChatCompletionMessageParam[] {
    const out: OpenAI.Chat.ChatCompletionMessageParam[] = [];
    if (opts?.system) out.push({ role: 'system', content: opts.system });
    for (const m of messages) out.push({ role: m.role, content: m.content });
    return out;
  }

  async complete(messages: LLMMessage[], opts?: CompletionOptions): Promise<CompletionResult> {
    try {
      const res = await this.client.chat.completions.create({
        model: opts?.model ?? this.defaultModel,
        ...(opts?.maxTokens ? { max_tokens: opts.maxTokens } : {}),
        messages: this.toMessages(messages, opts),
      });
      return {
        text: res.choices[0]?.message?.content ?? '',
        model: res.model,
        usage: {
          inputTokens: res.usage?.prompt_tokens ?? 0,
          outputTokens: res.usage?.completion_tokens ?? 0,
        },
      };
    } catch (cause) {
      throw new ProviderError('OpenAI completion failed', { cause });
    }
  }

  async *stream(messages: LLMMessage[], opts?: CompletionOptions): AsyncIterable<string> {
    try {
      const stream = await this.client.chat.completions.create({
        model: opts?.model ?? this.defaultModel,
        ...(opts?.maxTokens ? { max_tokens: opts.maxTokens } : {}),
        messages: this.toMessages(messages, opts),
        stream: true,
      });
      for await (const chunk of stream) {
        const delta = chunk.choices[0]?.delta?.content;
        if (delta) yield delta;
      }
    } catch (cause) {
      throw new ProviderError('OpenAI stream failed', { cause });
    }
  }

  async embed(input: string | string[], opts?: { model?: string }): Promise<number[][]> {
    try {
      const res = await this.client.embeddings.create({
        model: opts?.model ?? this.defaultEmbeddingModel,
        input,
      });
      return res.data.map((d) => d.embedding);
    } catch (cause) {
      throw new ProviderError('OpenAI embedding failed', { cause });
    }
  }
}

File: apps/voice/app/providers/adapters/openai.py
"""OpenAI streaming LLM — Python mirror of the TS `OpenAILLM` adapter.

Implements the `LLMProvider` contract over the Chat Completions API. Streaming uses the
documented SSE protocol (`data: {…}` lines, `[DONE]` sentinel — verified live) via the
httpx dependency we already have, so the voice loop gets first tokens fast without
pulling in another SDK. Cost is metered by the loop (tokens); the adapter never bills.
"""

from __future__ import annotations

import json
from collections.abc import AsyncIterator

import httpx

from app.providers.contracts import CompletionResult, LLMMessage, TokenUsage

API_URL = "https://api.openai.com/v1/chat/completions"


class LLMError(RuntimeError):
    """OpenAI completion failed (network or non-2xx)."""


class OpenAILLM:
    provider = "OPENAI"
    default_model = "gpt-4o-mini"

    def __init__(self, api_key: str) -> None:
        self._api_key = api_key

    def _payload(
        self, messages: list[LLMMessage], model: str | None, system: str | None
    ) -> dict[str, object]:
        wire: list[dict[str, str]] = []
        if system:
            wire.append({"role": "system", "content": system})
        wire.extend({"role": m.role, "content": m.content} for m in messages)
        return {"model": model or self.default_model, "messages": wire}

    @property
    def _headers(self) -> dict[str, str]:
        return {"Authorization": f"Bearer {self._api_key}", "Content-Type": "application/json"}

    async def complete(
        self,
        messages: list[LLMMessage],
        *,
        model: str | None = None,
        max_tokens: int | None = None,
        system: str | None = None,
    ) -> CompletionResult:
        payload = self._payload(messages, model, system)
        if max_tokens is not None:
            payload["max_tokens"] = max_tokens
        try:
            async with httpx.AsyncClient(timeout=httpx.Timeout(60.0, connect=10.0)) as client:
                res = await client.post(API_URL, headers=self._headers, json=payload)
                if res.status_code != 200:
                    raise LLMError(f"OpenAI completion error {res.status_code}: {res.text[:200]}")
                data = res.json()
        except httpx.HTTPError as exc:
            raise LLMError(f"OpenAI completion request failed: {exc}") from exc
        choice = data["choices"][0]["message"].get("content", "")
        usage = data.get("usage", {})
        return CompletionResult(
            text=choice,
            model=data.get("model", model or self.default_model),
            usage=TokenUsage(
                input_tokens=usage.get("prompt_tokens", 0),
                output_tokens=usage.get("completion_tokens", 0),
            ),
        )

    async def stream(
        self, messages: list[LLMMessage], *, model: str | None = None, system: str | None = None
    ) -> AsyncIterator[str]:
        payload = self._payload(messages, model, system)
        payload["stream"] = True
        try:
            async with (
                httpx.AsyncClient(timeout=httpx.Timeout(60.0, connect=10.0)) as client,
                client.stream("POST", API_URL, headers=self._headers, json=payload) as res,
            ):
                if res.status_code != 200:
                    detail = (await res.aread()).decode("utf-8", "replace")[:200]
                    raise LLMError(f"OpenAI stream error {res.status_code}: {detail}")
                async for line in res.aiter_lines():
                    if not line.startswith("data: "):
                        continue
                    data = line[6:]
                    if data == "[DONE]":
                        return
                    delta = json.loads(data)["choices"][0]["delta"].get("content")
                    if delta:
                        yield delta
        except httpx.HTTPError as exc:
            raise LLMError(f"OpenAI stream request failed: {exc}") from exc
